internal/pkg/instrumentation/probe/types.go

```diff
@@ -0,0 +1,324 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package probe provides instrumentation probe types and definitions.
+package probe
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"go.opentelemetry.io/auto/internal/pkg/inject"
+	"go.opentelemetry.io/auto/internal/pkg/instrumentation/bpffs"
+	"go.opentelemetry.io/auto/internal/pkg/instrumentation/probe/sampling"
+	"go.opentelemetry.io/auto/internal/pkg/instrumentation/utils"
+	"go.opentelemetry.io/auto/internal/pkg/process"
+	"go.opentelemetry.io/auto/internal/pkg/structfield"
+)
+
+// ErrInvalidConfig is intended to be used by a Probe's implementation of
+// ApplyConfig() when the provided probe.Config interface does not convert to
+// the Probe's custom config struct. For example, libraries may choose to ignore
+// this error when batch-updating Probes with a value that doesn't apply to all Probes.
+var ErrInvalidConfig = errors.New("invalid config type for Probe")
+
+const (
+	// The default size of the perf buffer in pages.
+	// We will need to make this configurable in the future.
+	PerfBufferDefaultSizeInPages = 128
+	// The default name of the eBPF map used to pass events from the eBPF program
+	// to userspace.
+	DefaultBufferMapName = "events"
+)
+
+// TracingConfig provides base configuration options for trace-based Probes.
+type TracingConfig struct {
+	SamplingConfig *sampling.Config
+}
+
+// TargetExecutableConfig provides executable and target process details for applications.
+type TargetExecutableConfig struct {
+	// Executable defines a target executable.
+	Executable *link.Executable
+	// TargetDetails defines target process information.
+	TargetDetails *process.TargetDetails
+}
+
+// BasicProbe is a provided implementation of BaseProbe. It is the base configuration
+// and identification layer for any Probe.
+type BasicProbe struct {
+	// ProbeID is a unique identifier for the Probe.
+	ProbeID ID
+	// Logger is used to log operations and errors.
+	Logger *slog.Logger
+}
+
+// ID returns the ID for this Probe.
+func (b *BasicProbe) ID() ID {
+	return b.ProbeID
+}
+
+// GetLogger returns the Logger for this Probe.
+func (b *BasicProbe) GetLogger() *slog.Logger {
+	return b.Logger
+}
+
+// TargetExecutableProbe is a provided implementation of GoLibraryTelemetryProbe.
+type TargetExecutableProbe[BPFObj any] struct {
```

Comment:
I wonder if we need the generic type here - as it seems we don't make use of it?